### DISCUSSION TITLE
Get Minimum Short Hash Length for Entire Repository

### DISCUSSION THREAD

**Discussion by rdamron1** [libgit2/libgit2#6442] [Q&A]
When you issue the "git log --format=%h" command, Git returns a short hash which is the minimum hash length for the entire repository. If you issue the same command against a shorter hash which is still unique, it will return the same longer length short hash.

Is there a way in LibGit2 to find the short hash length for the entire repository? 

I've tried calling git_object_short_id, but it returns the short hash with the minimum length for the target object which may be somewhat shorter than the minimum for repository which the git log command uses. And the set of git_oid_shorten_xxx functions are documented as returning the shortest hash length for the set of IDs being processed which would not be adequate either.

**Comment by ethomson** [comment]
Today there's no API to do this, though it would be straightforward to iterate all the objects in the repository to determine the minimum short hash length. A PR to provide an API would be welcome.

**Reply by rdamron1** [reply]
I'm not sure I understand what you mean by "A PR to provide an API would be welcome". 

Since the only abbreviation I can find on this site for PR is Pull Request, do you mean I should create the API and submit the code as a Pull Request? Or is "PR" the abbreviation for some kind of enhancement request? 

I have tried traversing the list of commits starting with the repository HEAD, but the length it obtains is always shorter than the length of a short hash returned by the git log command. How do you traverse all of the objects in the repository? What API is used to obtain the starting point, and what APIs are used to traverse the entire list?

**Comment by arroz** [comment]
Traversing all objects would be way too slow for this, especially since it would be something needed often (every time a hash is printed on screen?).

I looked into this a while ago and I believe core git estimates the minimum hash it needs based on the number of objects in the repository. It doesn't guarantee that, at that size, there will not be a conflict, but the probability of that happening is very low. And it's really fast since core git uses it to decide the size of the hashes it prints on every command output. Sadly I don't remember where exactly is that code in core git, but libgit2 should probably implement something similar.

**Reply by arroz** [reply]
Check https://github.com/git/git/blob/36f8e7ed7d72d2ac73743c3c2226cceb29b32156/object-name.c#L761

**Reply by arroz** [reply]
(It seems it actually *does* guarantee that since it disambiguates if there's a conflict after estimating the hash size?)

**Comment by ethomson** [comment]
> I have tried traversing the list of commits starting with the repository HEAD, but the length it obtains is always shorter than the length of a short hash returned by the git log command.

Interesting. I haven't examined this but I assume that it's returning the minimum of `core.abbrev` or the shortest hash length. 🤔

**Reply by rdamron1** [reply]
It's possible there are objects besides commits which require a longer short hash. 

I just need to know how to go about traversing all objects. This is something I would only do once during a process and keep it cached until the process ends, so I'm not worried about it being a bit slow.
